File: reorganize_data.py
import re

file_path = r'C:\Users\hirok\.gemini\antigravity\scratch\english-textbook-vocab\data.js'

def read_file(path):
    with open(path, 'r', encoding='utf-8') as f:
        return f.readlines()

def find_array_bounds(lines, key_line_pattern):
    start_index = -1
    for i, line in enumerate(lines):
        if re.search(key_line_pattern, line):
            start_index = i
            break
    
    if start_index == -1:
        return None, None

    # Find the opening bracket of the array
    array_open_index = -1
    for i in range(start_index, len(lines)):
        if '[' in lines[i]:
            array_open_index = i
            break
            
    if array_open_index == -1:
        return None, None
        
    # Find closing bracket
    balance = 0
    array_close_index = -1
    for i in range(array_open_index, len(lines)):
        line = lines[i]
        
        # Simple counting, might be fooled by comments/strings but usually ok for formatted code
        # We need to be careful not to count brackets in comments
        # Strip comments for counting
        clean_line = re.sub(r'//.*', '', line)
        clean_line = re.sub(r'/\*.*?\*/', '', clean_line) 
        
        balance += clean_line.count('[')
        balance -= clean_line.count(']')
        
        if balance == 0:
            array_close_index = i
            break
            
    return array_open_index, array_close_index

def parse_lessons(lines):
    lessons = []
    current_lesson = []
    balance = 0
    in_lesson = False
    
    for line in lines:
        stripped = line.strip()
        if not stripped:
            continue
            
        clean_line = re.sub(r'//.*', '', line)
        # simplistic comment removal
        
        if not in_lesson:
            if '{' in clean_line:
                in_lesson = True
                current_lesson.append(line)
                balance += clean_line.count('{')
                balance -= clean_line.count('}')
                if balance == 0:
                     lessons.append("".join(current_lesson))
                     current_lesson = []
                     in_lesson = False
        else:
            current_lesson.append(line)
            balance += clean_line.count('{')
            balance -= clean_line.count('}')
            if balance == 0:
                lessons.append("".join(current_lesson))
                current_lesson = []
                in_lesson = False
                
    return lessons

def get_lesson_info(lesson_str):
    unit_match = re.search(r'"unit":\s*"([^"]+)"', lesson_str)
    pages_match = re.search(r'"pages":\s*"([^"]+)"', lesson_str)
    words_match = re.search(r'"words":\s*\[(.*?)\]', lesson_str, re.DOTALL)
    
    unit = unit_match.group(1) if unit_match else ""
    pages = pages_match.group(1) if pages_match else ""
    has_words = bool(words_match and words_match.group(1).strip())
    
    return unit, pages, has_words

def process():
    lines = read_file(file_path)
    
    # Locate "new_crown" -> "2" and "3"
    # Assuming standard formatting as seen in view_file
    
    # Locate "2": [
    g2_start, g2_end = find_array_bounds(lines, r'"2":\s*\[')
    g3_start, g3_end = find_array_bounds(lines, r'"3":\s*\[')
    
    if g2_start is None or g3_start is None:
        print("Could not find Grade 2 or Grade 3 arrays.")
        return

    print(f"Grade 2 Array: Lines {g2_start+1}-{g2_end+1}")
    print(f"Grade 3 Array: Lines {g3_start+1}-{g3_end+1}")

    g2_content = lines[g2_start+1:g2_end]
    g3_content = lines[g3_start+1:g3_end]
    
    g2_lessons_raw = parse_lessons(g2_content)
    g3_lessons_raw = parse_lessons(g3_content)
    
    all_lessons = []
    
    # Add existing Grade 3 lessons
    for l in g3_lessons_raw:
        # cleanup comma at end if present
        clean_l = l.strip().rstrip(',')
        all_lessons.append(clean_l)
        
    # Process Grade 2 lessons to find misplaced Grade 3 ones
    for l in g2_lessons_raw:
        unit, pages, has_words = get_lesson_info(l)
        clean_l = l.strip().rstrip(',')
        
        print(f"Found in Grade 2: {unit} ({pages}) Words: {has_words}")
        
        # Filter Logic:
        # We want Lessons 4, 5, 6, 7, 8 for Grade 3.
        # Check for empty Lesson 5 (P57-68) -> Skip
        # Check for duplicate Lesson 6 (P69-79) -> Skip if we have better one?
        # Actually, let's look at the specific attributes
        
        if unit == "Lesson 4":
            all_lessons.append(clean_l)
        elif unit == "Lesson 5":
            if "P57〜68" in pages and not has_words:
                print("Skipping empty Lesson 5")
                continue
            all_lessons.append(clean_l)
        elif unit == "Lesson 6":
            if "P69〜79" in pages:
                 print("Skipping partial/old Lesson 6 (P69-79)")
                 continue
            all_lessons.append(clean_l)
        elif unit == "Lesson 7":
            all_lessons.append(clean_l)
        elif unit == "Lesson 8":
            all_lessons.append(clean_l)
        else:
            print(f"Unknown unit in Grade 2: {unit}. Keeping it? No, assume it's garbage or old G2.")
            # If it's real Grade 2 (unlikely based on my analysis), we might lose it. 
            # But the prompt implies I need to fix the mess I made.
            pass

    # Sort lessons by Lesson number just in case
    # Lesson 1, Lesson 2, ...
    def sort_key(lesson_str):
        unit, _, _ = get_lesson_info(lesson_str)
        num = re.search(r'\d+', unit)
        return int(num.group(0)) if num else 999
        
    all_lessons.sort(key=sort_key)
    
    # Reconstruct the file
    # We will clear Grade 2 and populate Grade 3
    
    new_g3_content = "\n" + ",\n".join(all_lessons) + "\n"
    
    # We reconstruct the file by parts
    # Part 1: Start to g2_start
    # Part 2: "2": [],
    # Part 3: g2_end+1 to g3_start+1
    # Part 4: content of g3
    # Part 5: g3_end to end
    
    # Wait, indices are tricky.
    # header: lines[:g2_start+1] -> includes "2": [
    # But we want to empty it.
    
    # Let's verify overlap. G2 comes before G3.
    
    final_lines = []
    final_lines.extend(lines[:g2_start+1]) # Up to "2": [
    final_lines.append("\n        ],\n") # Empty Grade 2
    
    # Middle part: After G2 end, up to G3 start
    # g2_end is the line with "]"
    
    # If there is stuff between g2_end and g3_start?
    # Usually: "2": [ ... ], \n "3": [
    
    final_lines.extend(lines[g2_end+1 : g3_start+1]) # Includes "3": [
    
    final_lines.append(new_g3_content)
    
    final_lines.extend(lines[g3_end:]) # Closing bracket of G3 and rest
    
    with open(file_path, 'w', encoding='utf-8') as f:
        f.writelines(final_lines)
    
    print("Successfully reorganized data.js")

if __name__ == '__main__':
    try:
        process()
    except Exception as e:
        import traceback
        traceback.print_exc()

File: reorganize_data.js
const fs = require('fs');
const path = require('path');

const filePath = path.join(__dirname, 'data.js');
const source = fs.readFileSync(filePath, 'utf8');
const lines = source.split('\n');

function findArrayBounds(keyPattern) {
    const startIdx = lines.findIndex(line => keyPattern.test(line));
    if (startIdx === -1) return null;

    // Find opening bracket
    let arrayOpenIdx = -1;
    for (let i = startIdx; i < lines.length; i++) {
        if (lines[i].includes('[')) {
            arrayOpenIdx = i;
            break;
        }
    }
    if (arrayOpenIdx === -1) return null;

    let balance = 0;
    let arrayCloseIdx = -1;

    for (let i = arrayOpenIdx; i < lines.length; i++) {
        let line = lines[i];
        // Strip comments for bracket counting
        line = line.replace(/\/\/.*$/, '').replace(/\/\*.*?\*\//g, '');

        for (const char of line) {
            if (char === '[') balance++;
            if (char === ']') balance--;
        }

        if (balance === 0) {
            arrayCloseIdx = i;
            break;
        }
    }

    return { start: arrayOpenIdx, end: arrayCloseIdx, headerStart: startIdx };
}

const g2 = findArrayBounds(/"2":\s*\[/);
const g3 = findArrayBounds(/"3":\s*\[/);

if (!g2 || !g3) {
    console.error('Could not find Grade 2 or Grade 3 arrays');
    process.exit(1);
}

console.log(`Grade 2: ${g2.start}-${g2.end}`);
console.log(`Grade 3: ${g3.start}-${g3.end}`);

const g2Content = lines.slice(g2.start + 1, g2.end).join('\n');
const g3Content = lines.slice(g3.start + 1, g3.end).join('\n');

function parseLessons(text) {
    const rawLessons = [];
    let buffer = '';
    let balance = 0;
    let inLesson = false;

    const contentLines = text.split('\n');

    for (const line of contentLines) {
        const cleanLine = line.replace(/\/\/.*$/, '').replace(/\/\*.*?\*\//g, '');

        // Very simplified parsing logic assuming standard indentation
        // We look for { at start of an object and } at end

        if (cleanLine.trim() === '') continue;

        if (!inLesson) {
            if (cleanLine.includes('{')) {
                inLesson = true;
                buffer += line + '\n';
                for (const char of cleanLine) {
                    if (char === '{') balance++;
                    if (char === '}') balance--;
                }
                if (balance === 0) {
                    rawLessons.push(buffer);
                    buffer = '';
                    inLesson = false;
                }
            }
        } else {
            buffer += line + '\n';
            for (const char of cleanLine) {
                if (char === '{') balance++;
                if (char === '}') balance--;
            }
            if (balance === 0) {
                rawLessons.push(buffer);
                buffer = '';
                inLesson = false;
            }
        }
    }
    return rawLessons;
}

const g2Lessons = parseLessons(g2Content);
const g3Lessons = parseLessons(g3Content);

console.log(`Found ${g2Lessons.length} lessons in Grade 2`);
console.log(`Found ${g3Lessons.length} lessons in Grade 3`);

function analyzeLesson(lessonStr) {
    const unitM = lessonStr.match(/"unit":\s*"([^"]+)"/);
    const pagesM = lessonStr.match(/"pages":\s*"([^"]+)"/);

    // Check if word list is non-empty
    // Simple heuristic: look for "en":
    const hasWords = lessonStr.includes('"en":');

    return {
        unit: unitM ? unitM[1] : '',
        pages: pagesM ? pagesM[1] : '',
        hasWords: hasWords,
        fullText: lessonStr.trim().replace(/,$/, '')
    };
}

const allLessons = [];

// Keep existing G3
g3Lessons.forEach(l => {
    const info = analyzeLesson(l);
    if (info.unit) allLessons.push(info);
});

// Move G2 to G3
g2Lessons.forEach(l => {
    const info = analyzeLesson(l);
    console.log(`Evaluating G2 lesson: ${info.unit} ${info.pages} Words:${info.hasWords}`);

    if (info.unit === 'Lesson 4') allLessons.push(info);
    else if (info.unit === 'Lesson 5') {
        if (!info.hasWords) {
            console.log('Skipping empty Lesson 5');
        } else {
            allLessons.push(info);
        }
    }
    else if (info.unit === 'Lesson 6') {
        if (info.pages.includes('P69〜79')) {
            console.log('Skipping partial Lesson 6');
        } else {
            allLessons.push(info);
        }
    }
    else if (info.unit === 'Lesson 7') allLessons.push(info);
    else if (info.unit === 'Lesson 8') allLessons.push(info);
});

allLessons.sort((a, b) => {
    const getNum = (u) => {
        const m = u.match(/\d+/);
        return m ? parseInt(m[0]) : 999;
    };
    return getNum(a.unit) - getNum(b.unit);
});

const newG3Str = '\n' + allLessons.map(l => l.fullText).join(',\n') + '\n';

const finalLines = [
    ...lines.slice(0, g2.start + 1),
    ...lines.slice(g2.end, g3.start + 1),
    newG3Str,
    ...lines.slice(g3.end)
];

fs.writeFileSync(filePath, finalLines.join('\n'));
console.log('Success Reorganizing');
